plugins/backup/veeam/src/main/java/org/apache/cloudstack/backup/VeeamBackupProvider.java

```diff
@@ -0,0 +1,319 @@
+        final VmwareDatacenter vmwareDatacenter = findVmwareDatacenterForVM(vm);
+        return getClient(vm.getDataCenterId()).removeVMFromVeeamJob(policy.getExternalId(), vm.getInstanceName(), vmwareDatacenter.getVcenterHost());
+    }
+
+    private String getGuestBackupName(final String instanceName, final String uuid) {
+        return String.format("%s%s%s", instanceName, BACKUP_IDENTIFIER, uuid);
+    }
+
+    @Override
+    public boolean assignVMToBackupOffering(final VirtualMachine vm, final BackupOffering backupOffering) {
+        final VeeamClient client = getClient(vm.getDataCenterId());
+        final Job parentJob = client.listJob(backupOffering.getExternalId());
+        final String clonedJobName = getGuestBackupName(vm.getInstanceName(), vm.getUuid());
+        if (client.cloneVeeamJob(parentJob, clonedJobName)) {
+            for (BackupOffering job : client.listJobs()) {
+                if (job.getName().equals(clonedJobName)) {
+                    final Job clonedJob = client.listJob(job.getExternalId());
+                    if (clonedJob.getScheduleConfigured() && !clonedJob.getScheduleEnabled()) {
+                        client.toggleJobSchedule(clonedJob.getId());
+                    }
+                    final VmwareDatacenter vmwareDC = findVmwareDatacenterForVM(vm);
+                    if (client.addVMToVeeamJob(job.getExternalId(), vm.getInstanceName(), vmwareDC.getVcenterHost())) {
+                        ((VMInstanceVO) vm).setBackupExternalId(job.getExternalId());
+                        return true;
+                    }
+                }
+            }
+        } else {
+            LOG.error("Failed to clone pre-defined Veeam job (backup offering) for backup offering ID: " + backupOffering.getExternalId());
+        }
+        return false;
+    }
+
+    @Override
+    public boolean removeVMFromBackupOffering(final VirtualMachine vm) {
+        final VeeamClient client = getClient(vm.getDataCenterId());
+        final VmwareDatacenter vmwareDC = findVmwareDatacenterForVM(vm);
+        try {
+            if (!client.removeVMFromVeeamJob(vm.getBackupExternalId(), vm.getInstanceName(), vmwareDC.getVcenterHost())) {
+                LOG.warn("Failed to remove VM from Veeam Job id: " + vm.getBackupExternalId());
+            }
+        } catch (CloudRuntimeException e) {
+            LOG.debug("VM was removed from the job so could not remove again, trying to delete the veeam job now.", e);
+        }
+
+        final String clonedJobName = getGuestBackupName(vm.getInstanceName(), vm.getUuid());
+        if (!client.deleteJobAndBackup(clonedJobName)) {
+            LOG.warn("Failed to remove Veeam job and backup for job: " + clonedJobName);
+            throw new CloudRuntimeException("Failed to delete Veeam B&R job and backup, an operation may be in progress. Please try again after some time.");
+        }
+        for (final Backup backup: backupDao.listByVmId(null, vm.getId())) {
```

Comment:
Can this also be moved to the BackupManagerImpl on successful operation?

Reply:
Good idea, done.